[Site Isolation] Enter fullscreen animation begin needs coordinate transformation with site isolation on

nipunshukla21 opened the pull request:
#### 10bc7cf26356bbc80c3261052768d5c8ff8d7fc4
<pre>
[Site Isolation] Enter fullscreen animation begin needs coordinate transformation with site isolation on
<a href="https://bugs.webkit.org/show_bug.cgi?id=297673">https://bugs.webkit.org/show_bug.cgi?id=297673</a>
<a href="https://rdar.apple.com/153776866">rdar://153776866</a>

Reviewed by Alex Christensen and Andy Estes.

Previously, when attempting to full screen a cross origin video with site isolation
enabled, the animation would originate from the top left corner of the window.
This is because all transformations were done in the child frame and it was unable
to determine its position within the main frame&apos;s coordinate space, instead defaulting
to (0, 0). This change modifies IPC for the child frame to communicate with the
WebFullScreenManagerProxy in the UIProcess, which then recurses up the process boundary
to determine the coordinates in the main frame before sending those back to the UIProcess,
which converts them to screen space coordinates for use in the animation.

* LayoutTests/http/tests/site-isolation/fullscreen-expected.txt:
* LayoutTests/http/tests/site-isolation/fullscreen.html:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::enterFullScreen):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.messages.in:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::rootViewRectOfContents):
(WebKit::getRootFrameCoordinatesAndIdentifier):
(WebKit::WebFullScreenManager::enterFullScreenForElement):
(WebKit::WebFullScreenManager::willEnterFullScreen):
(WebKit::WebFullScreenManager::willExitFullScreen):
(WebKit::screenRectOfContents):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::beganEnterFullScreen):
(WTR::TestController::beganExitFullScreen):
(WTR::TestController::resetStateToConsistentValues):
* Tools/WebKitTestRunner/TestController.h:
(WTR::TestController::dumpFullScreenOrigin):

Canonical link: <a href="https://commits.webkit.org/301065@main">https://commits.webkit.org/301065@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/09803232e4f394bad58da11873df52f90bdb1c24

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124539 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44215 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34950 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131374 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76505 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126416 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44927 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52793 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94740 "Found 1 new test failure: http/wpt/service-workers/online.https.html (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62828 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127493 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35811 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111377 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75316 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34749 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29532 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74856 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105565 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29764 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134042 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51401 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39225 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103219 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51810 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107600 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103001 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26289 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48364 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26626 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/48341 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51273 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57065 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50681 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54037 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52362 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -10124,6 +10124,37 @@ void WebPage::remoteDictionaryPopupInfoToRootView(WebCore::FrameIdentifier frame
+    if (!mainFrame) {
+        completionHandler(childFrameRect);
+        return;
+    }
```
I like the slightly more concise syntax like this:
if (!mainFrame)
    return completionHandler(childFrameRect);

achristensen07 commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -489,19 +548,28 @@ void WebFullScreenManager::willExitFullScreen(CompletionHandler<void()>&& comple
-    }
+    auto proceedWithExitFullScreen = [this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler)] (WebCore::IntRect finalFrame) mutable {
+        m_finalFrame = finalFrame;
+        if (!m_element->document().fullscreen().willExitFullscreen()) {
```
We should probably null check m_element, which may have become null by the time this lambda is called.

achristensen07 commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -269,6 +269,46 @@ FullScreenMediaDetails WebFullScreenManager::getImageMediaDetails(CheckedPtr<Ren
 }
 #endif // ENABLE(QUICKLOOK_FULLSCREEN)
 
+bool WebFullScreenManager::needsAsyncCoordinateTransformation(WebCore::Element& element)
```
I don't think this function is quite right.  We need an asynchronous coordinate transformation if any frame in the ancestor chain is remote, not just if the main one is remote.  Imagine we are in a grandchild iframe in the same domain as the grandparent main frame but not the parent iframe.

nipunshukla21 replied:
That makes sense. I've changed it to return true if any frame in the chain up is remote.

achristensen07 commented on LayoutTests/http/tests/site-isolation/fullscreen-expected.txt:
```diff
@@ -1,8 +1,10 @@
 supportsFullScreen() == true
 enterFullScreenForElement()
 beganEnterFullScreen() - initialRect.size: {300, 150}, finalRect.size: {800, 600}
+beganEnterFullScreen() - initialRect.origin: {8, 442}, finalRect.origin: {-8, -442}
```
Are these coordinates wrong before your fix?

nipunshukla21 replied:
Yes, they're consistently (0, 450) without my patch -- although now I see that even if the coordinates are different in the dump vs the -expected.txt file it doesn't trigger a test failure but does update the txt if I delete it and recreate expectations by running the test... I'm investigating that now.

EDIT: Interesting that in LayoutTest results it shows that there's a diff when the test runs incorrectly with my fix disabled but the `run-webkit-tests` says it passed:
```❯ cat fullscreen-diff.txt
--- /Users/nipunshukla/OpenSource/WebKitBuild/Release/layout-test-results/http/tests/site-isolation/fullscreen-expected.txt
+++ /Users/nipunshukla/OpenSource/WebKitBuild/Release/layout-test-results/http/tests/site-isolation/fullscreen-actual.txt
@@ -1,10 +1,10 @@
 supportsFullScreen() == true
 enterFullScreenForElement()
 beganEnterFullScreen() - initialRect.size: {300, 150}, finalRect.size: {800, 600}
-beganEnterFullScreen() - initialRect.origin: {8, 442}, finalRect.origin: {-8, -442}
+beganEnterFullScreen() - initialRect.origin: {0, 450}, finalRect.origin: {0, -450}
 exitFullScreenForElement()
 beganExitFullScreen() - initialRect.size: {800, 600}, finalRect.size: {300, 150}
-beganExitFullScreen() - initialRect.origin: {-8, -442}, finalRect.origin: {8, 442}
+beganExitFullScreen() - initialRect.origin: {0, -450}, finalRect.origin: {0, 450}

 Size after entering fullscreen: 600x800
 iframe border style after transition: 0px none rgb(0, 0, 0)
 ```
 I can't tell what's happening here. Maybe I need to clean build or something?

achristensen07 reviewed (changes requested):
I believe this works, but I don't understand conceptually why the web content process needs to know the transformed start coordinates.  Shouldn't it be working only with relative coordinates?

achristensen07 commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -339,14 +354,15 @@ void WebFullScreenManager::enterFullScreenForElement(Element& element, HTMLMedia
-            if (success) {
+        ] (std::optional<WebCore::IntRect> transformedInitialFrame) mutable {
+            if (transformedInitialFrame) {
+                m_initialFrame = *transformedInitialFrame;
```
Ideally this would replace line 321 where we set m_initialFrame to screenRectOfContents(element).  Also, we should probably rename the variable to make clear what coordinates the frame is in.

achristensen07 commented on Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
```diff
@@ -212,17 +212,28 @@ Awaitable<bool> WebFullScreenManagerProxy::enterFullScreen(IPC::Connection& conn
+    if (RefPtr page = m_page.get()) {
+        if (RefPtr mainFrame = page->mainFrame()) {
+            if (RefPtr mainFrameProcess = mainFrame->process()) {
+                std::optional<FloatRect> mainFrameCoordinates = co_await page->convertRectToMainFrameCoordinates(WebCore::FloatRect(initialFrame), frameID);
```
In order for this to be 0 IPC roundtrips in the non-site-isolation case, we would want to pass in the root frame identifier and the coordinates in root frame coordinates.  Root frame is the topmost local frame in a frame tree subtree in a given process.

Also, this can be written with a lot less indentation.

achristensen07 commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -69,7 +69,12 @@ using namespace WebCore;
 using WebCore::FloatSize;
 
-static WebCore::IntRect screenRectOfContents(WebCore::Element& element)
+enum class CoordinateConversionType {
```
: bool
There are only 2 values.

achristensen07 commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -93,7 +98,17 @@ static WebCore::IntRect screenRectOfContents(WebCore::Element& element)
+    return type == CoordinateConversionType::Screen ? frameView->contentsToScreen(contentsRect) : frameView->contentsToRootView(contentsRect);
+}
+
+static WebCore::IntRect screenRectOfContents(WebCore::Element& element)
```
I think when this PR is done we won't be doing any conversions to screen coordinates in this file.

achristensen07 commented on Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
```diff
@@ -58,6 +58,7 @@ interface TestRunner {
     undefined dumpEditingCallbacks();
     undefined dumpSelectionRect();
     undefined dumpTitleChanges();
+    undefined dumpFullScreenOrigin();
```
Can this be in testRunnerJS instead of TestRunner.idl?  That's the direction we're heading.

aestes commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -93,7 +94,16 @@ static WebCore::IntRect screenRectOfContents(WebCore::Element& element)
+    CheckedPtr renderer = element.renderer();
+    if (!renderer)
+        return;
+    frameID = renderer->protectedFrame()->rootFrame().frameID();
```
Nit: could be a little more compact
```suggestion
    if (CheckedPtr renderer = element.renderer())
        frameID = renderer->protectedFrame()->rootFrame().frameID();
```

nipunshukla21 committed:
[Site Isolation] Enter fullscreen animation begin needs coordinate transformation with site isolation on
https://bugs.webkit.org/show_bug.cgi?id=297673
rdar://153776866

Reviewed by Alex Christensen and Andy Estes.

Previously, when attempting to full screen a cross origin video with site isolation
enabled, the animation would originate from the top left corner of the window.
This is because all transformations were done in the child frame and it was unable
to determine its position within the main frame's coordinate space, instead defaulting
to (0, 0). This change modifies IPC for the child frame to communicate with the
WebFullScreenManagerProxy in the UIProcess, which then recurses up the process boundary
to determine the coordinates in the main frame before sending those back to the UIProcess,
which converts them to screen space coordinates for use in the animation.

* LayoutTests/http/tests/site-isolation/fullscreen-expected.txt:
* LayoutTests/http/tests/site-isolation/fullscreen.html:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::enterFullScreen):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.messages.in:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::rootViewRectOfContents):
(WebKit::getRootFrameCoordinatesAndIdentifier):
(WebKit::WebFullScreenManager::enterFullScreenForElement):
(WebKit::WebFullScreenManager::willEnterFullScreen):
(WebKit::WebFullScreenManager::willExitFullScreen):
(WebKit::screenRectOfContents):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::beganEnterFullScreen):
(WTR::TestController::beganExitFullScreen):
(WTR::TestController::resetStateToConsistentValues):
* Tools/WebKitTestRunner/TestController.h:
(WTR::TestController::dumpFullScreenOrigin):

Canonical link: https://commits.webkit.org/301065@main
